blockchain/pallets/token_distribution/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+            address: Destination<T::AccountId>,
+            #[pallet::compact] weight: u32,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            DestinationWeights::<T>::insert(address, weight);
+            Ok(())
+        }
+
+        #[pallet::weight((
+            10_000 + T::DbWeight::get().reads_writes(1, 1),
+            DispatchClass::Normal,
+            Pays::No
+        ))]
+        pub fn increment_issuance_offset(
+            origin: OriginFor<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            TotalIssuanceOffset::<T>::set(TotalIssuanceOffset::<T>::get() + amount);
+            Ok(())
+        }
+
+        #[pallet::weight((
+            10_000 + T::DbWeight::get().reads_writes(1, 1),
+            DispatchClass::Normal,
+            Pays::No
+        ))]
+        pub fn mint(
+            origin: OriginFor<T>,
+            address: T::AccountId,
+            #[pallet::compact] amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            T::Currency::deposit_creating(&address, amount);
+            TotalIssuanceOffset::<T>::set(TotalIssuanceOffset::<T>::get() + amount);
+
+            Ok(())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn take_from(purpose: u8) -> BalanceOf<T> {
+            PurposeBalances::<T>::take(purpose)
+        }
+
+        pub fn return_to(purpose: u8, amount: BalanceOf<T>) {
+            PurposeBalances::<T>::insert(purpose, amount);
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+    where
+        BalanceOf<T>: num_traits::PrimInt + core::iter::Sum,
+        BlockNumberFor<T>: num_traits::PrimInt,
+    {
+        fn on_finalize(block_number: BlockNumberFor<T>) {
+            let total_weight: u32 = DestinationWeights::<T>::iter_values().sum();
+            if total_weight == 0 {
+                return;
+            }
+
+            let already_issued = T::Currency::total_issuance()
+                + PurposeBalances::<T>::iter_values().sum()
+                - TotalIssuanceOffset::<T>::get();
+
+            // Using Issuance like this makes it _technically_ possible for
+            // consensus to fail if the CPU's floating point calculations are
+            // different.
+            //
+            // If this becomes a problem, we can have this value derived from
+            // an extrinsic that an authoritative account sets. Similar to how
+            // the timestamp pallet works.
+            let issuance = crate::Issuance {
+                total: T::TotalIssuance::get(),
+                half_life: T::IssuanceHalfLife::get(),
+                complete_at: T::IssuanceCompleteAt::get(),
+            };
+            let to_issue = issuance.total_issued_by(block_number) - already_issued;
+            let unit_balance = to_issue / total_weight.into();
```

Comment:
Could this be an issue (pun not intended) at the very end, when `to_issue` is smaller than `total_weight`?